discovery-provider/src/challenges/challenge_event_bus.py

```diff
@@ -55,6 +55,32 @@ def dispatch(
         except Exception as e:
             logger.warning(f"ChallengeEventBus: error enqueuing to Redis: {e}")
 
+    def queue_event(
+        self,
+        event: str,
+        block_number: int,
+        user_id: int,
+        extra: Dict = None,
+    ):
+        """Queues dispatching an event + block_number + user_id to Redis queue"""
+        self._event_queue.append(
+            {
+                "event": event,
+                "block_number": block_number,
+                "user_id": user_id,
+                "extra": extra,
+            }
+        )
+
+    def flush_event_queue(self):
+        for event in self._event_queue:
+            self.dispatch(
+                event["event"],
+                event["block_number"],
+                event["user_id"],
+                event["extra"] if event["extra"] else {},
+            )
```

Comment:
We need to dump the queue after we flush to redis! And will have to be careful that we only dump things we added to redis at the start of this method call, since this queue is being accessed on multiple threads now